### PR TITLE
Feature/ldap support

### PR DESCRIPTION
**Pull request by coutoPL**
LDAP support for readonlyrest plugin.

- to run tests docker is required
- sadly, I disabled integration `ConfigurationTests` because I had a problem with disappearing transitive dependencies of `testcontainers-java`. I have no idea, how to make it works with `org.elasticsearch.test` lib (have no time to investigate it, sorry).

### CODE REVIEW

**Comment by sscarduzio** [comment]
@coutoPL in the meanwhile I prepare the code to the merge, could you please add the license headers in all the new Java and property files?

**Comment by coutoPL** [comment]
yes, I'll do it today evening. I see, there is problem with travis build. `./gradle assembly` seems to work, but `./gradlew updateSHAs check assemble` ends up with jarhell error :| Will try to fix it too.

**Comment by sscarduzio** [comment]
@coutoPL could you also address those easy fixes from @codacy-bot? As soon as you're ready I'm merging :)

**Comment by coutoPL** [comment]
Yes, I'm currently working on it.

**Comment by coutoPL** [comment]
OK, ready. But in build script I had to remove `check` task because jarhell returns "jar hell!" issue with `guava` and `testcontainers-java`. Sadlu, there is no option to disable jarhell check in tests only :(

File: src/main/java/org/elasticsearch/plugin/readonlyrest/ldap/LdapClientException.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *    GNU General Public License for more details.
+ *
+ *    You should have received a copy of the GNU General Public License
+ *    along with ReadonlyREST.  If not, see http://www.gnu.org/licenses/
+ */
+
+package org.elasticsearch.plugin.readonlyrest.ldap;
+
+public class LdapClientException {
```
Comment (coutoPL) [inline src/main/java/org/elasticsearch/plugin/readonlyrest/ldap/LdapClientException.java]: won't fix

File: src/main/java/org/elasticsearch/plugin/readonlyrest/ldap/UnboundidLdapClient.java [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+    public static Duration DEFAULT_LDAP_CACHE_TTL = Duration.ZERO;
+    public static boolean DEFAULT_LDAP_SSL_ENABLED = true;
+    public static boolean DEFAULT_LDAP_SSL_TRUST_ALL_CERTS = false;
+
+    private final String searchUserBaseDN;
+    private final String searchGroupBaseDN;
+    private final LDAPConnectionPool connectionPool;
+    private final Long timeout;
+
+    private UnboundidLdapClient(String host,
```
Comment (coutoPL) [inline src/main/java/org/elasticsearch/plugin/readonlyrest/ldap/UnboundidLdapClient.java]: won't fix

File: src/main/java/org/elasticsearch/plugin/readonlyrest/IndexLevelActionFilter.java [hunk 3]
```diff
@@ -1,131 +1,146 @@
...
+    public int order() {
+        return 0;
     }
-    boolean reqNull = req == null;
 
-    // This was not a REST message
-    if (reqNull && chanNull) {
-      return true;
+    @Override
+    public <Request extends ActionRequest, Response extends ActionResponse> void apply(Task task,
```
Comment (coutoPL) [inline src/main/java/org/elasticsearch/plugin/readonlyrest/IndexLevelActionFilter.java]: won't fix

File: src/main/java/org/elasticsearch/plugin/readonlyrest/acl/blocks/Block.java [hunk 4]
```diff
@@ -23,178 +22,244 @@
...
-
-    public static String valuesString() {
-      StringBuilder sb = new StringBuilder();
-      for (Policy v : values()) {
-        sb.append(v.toString()).append(",");
-      }
-      sb.deleteCharAt(sb.length() - 1);
-      return sb.toString();
+
+    private void initSyncConditions(Settings s, List<User> userList) {
```
Comment (coutoPL) [inline src/main/java/org/elasticsearch/plugin/readonlyrest/acl/blocks/Block.java]: won't fix

**Comment by sscarduzio** [comment]
You mentioned I can test this with docker, can you expand?

**Comment by coutoPL** [comment]
Two tests (`LdapConfigTests` and `UnboundidLdapClientTests` classes) are using ldap docker image. Before test suite docker container is created and configured (see `LdapContainer` class), then all tests in the suite are executed and docker container is being removed. So, if docker daemon in your environment (I've also added proper settings to travis conf) is available, tests should end up with success. Just invoke `./gradle clean test`.

**Comment by sscarduzio** [comment]
cool, but I can't find those classes in the list of changed files in the PR

**Comment by sscarduzio** [comment]
OK Now I found them.. wtf is wrong with me.

**Comment by sscarduzio** [comment]
BTW I love testcontainers, didn't know it existed!

**Comment by coutoPL** [comment]
thanks for fast merge!

**Pull request merged**